### PR TITLE
[FEATURE] Introducing General-Purpose Wrapper for Regular Expressions Parsing and incorporating it in "_FilesystemDataAsset"

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* A new wrapper class, `RegExParser`, on top of `Python` `re` package is introduced for matching purposes and incorporated in `_FilesystemDataAsset`.  As a result, the usage of Regular Expressions in `_FilesystemDataAsset` is simpler, easier to understand, and exhibits better performance, due to not repeating parsing operations unnecessarily.
* Thorough unit tests of `RegExParser` are included as well.
* Stylistic cleanup is also provided.

### Next TODO
This new general `RegExParser` module will be instrumental to utilization of `DataConnector` in ZEP `DataAsset` implementations for all "file-path" type of data resources (i.e., local filesystem, S3, GCS, Azure, etc.).

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-579/GREAT-1646
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/experimental/datasources/filesystem_data_asset.py [hunk 1]
```diff
@@ -28,7 +31,7 @@
 
 class _FilesystemDataAsset(DataAsset):
     # Pandas specific class attrs
-    _EXCLUDE_FROM_READER_OPTIONS: ClassVar[Set[str]] = {
+    _EXCLUDE_FROM_READER_OPTIONS: ClassVar[set[str]] = {
```
Comment (Kilo59) [inline great_expectations/experimental/datasources/filesystem_data_asset.py]: You need to revert this

File: great_expectations/experimental/datasources/pandas_datasource.py [hunk 2]
```diff
@@ -152,9 +152,9 @@ def add_excel_asset(
         """
         asset = ExcelAsset(
             name=name,
-            base_directory=base_directory,  # type: ignore[arg-type]  # str will be coerced to Path
-            regex=regex,  # type: ignore[arg-type]  # str with will coerced to Pattern
-            order_by=order_by or [],  # type: ignore[arg-type]  # coerce list[str]
```
Comment (Kilo59) [inline great_expectations/experimental/datasources/pandas_datasource.py:157]: I'm confused as to why these aren't type errors anymore??

Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/pandas_datasource.py:157]: @Kilo59 I do not know.  But Azure complained that these "ignores" are no longer needed.  I mentioned this in Slack.

File: great_expectations/experimental/datasources/filesystem_data_asset.py [hunk 3]
```diff
@@ -55,13 +63,28 @@ class Config:
...
+            unnamed_regex_group_prefix=self._unnamed_regex_param_prefix,
+        )
+
+        self._all_group_name_to_group_index_mapping = (
+            self._regex_parser.get_all_group_name_to_group_index_mapping()
+        )
+        self._all_group_index_to_group_name_mapping = (
+            self._regex_parser.get_all_group_index_to_group_name_mapping()
+        )
+        self._all_group_names = self._regex_parser.get_all_group_names()
```
Comment (Kilo59) [inline great_expectations/experimental/datasources/filesystem_data_asset.py:79]: I think you could do all these with `validator` methods, and then you wouldn't need this custom `__init__`.
https://docs.pydantic.dev/usage/validators/

🤔 
I think this is why the CI mypy said those ignores were no longer needed.

Will update this comment with an example implementation.

Comment (Kilo59) [reply great_expectations/experimental/datasources/filesystem_data_asset.py:79]: Actually, I don't think we can do these with `pydantic.PrivateAttr()`.
I think we should make these normal attributes and then we can get rid of the custom init but I don't want to block the PR on this.

I'll followup with you and maybe we can pair on this.

**Review by Kilo59 — APPROVED** [review]
I'd like to followup with you on updating `RegExParser` so that it can tell pydantic how it should be validated (with a `__get_validators__` method).
Then we can get rid of the custom `__init__` on `_FileDataAsset` and simplify the code.

https://docs.pydantic.dev/usage/types/#custom-data-types

**Pull request merged**